<a href="https://colab.research.google.com/github/Panchol211/Week-4-IP/blob/master/Moringa_Data_Science_Prep_W4_Independent_Project_2020_07_Panchol_Alier_NoteBook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
## This is the analysis of the data from Elctrica car company in France 
import pandas as pd
import numpy as np


In [44]:
df = pd.read_csv("Autolib_dataset.csv")
df.head(10)

Address  Cars  Bluecar counter  ...  day  hour  minute
0                 2 Avenue de Suffren     0                0  ...    8    11      43
1           145 Rue Raymond Losserand     6                6  ...    6     7      24
2    2 Avenue John Fitzgerald Kennedy     3                3  ...    3    20      14
3              51 Rue Eugène Oudiné     3                3  ...    4     4      37
4  6 avenue de la Porte de Champerret     3                3  ...    8    17      23
5                8 Boulevard Voltaire     0                0  ...    6     7       2
6                      37 rue Leblanc     0                0  ...    8    18      20
7         17 Rue des Luaps Prolongée     3                3  ...    4    22      13
8               34 avenue Jean Moulin     1                1  ...    2    22      58
9        41 boulevard de Rochechouart     6                6  ...    4    15       2

[10 rows x 25 columns]

In [ ]:
# Trying to understand the data there are about 5000 rows and 25 columns 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Address              5000 non-null   object
 1   Cars                 5000 non-null   int64 
 2   Bluecar counter      5000 non-null   int64 
 3   Utilib counter       5000 non-null   int64 
 4   Utilib 1.4 counter   5000 non-null   int64 
 5   Charge Slots         5000 non-null   int64 
 6   Charging Status      5000 non-null   object
 7   City                 5000 non-null   object
 8   Displayed comment    111 non-null    object
 9   ID                   5000 non-null   object
 10  Kind                 5000 non-null   object
 11  Geo point            5000 non-null   object
 12  Postal code          5000 non-null   int64 
 13  Public name          5000 non-null   object
 14  Rental status        5000 non-null   object
 15  Scheduled at         47 non-null     object
 16  Slots 

In [ ]:
## checking for missing values and based on the out put there is not missing values 
df.isnull().count()

Address                5000
Cars                   5000
Bluecar counter        5000
Utilib counter         5000
Utilib 1.4 counter     5000
Charge Slots           5000
Charging Status        5000
City                   5000
Displayed comment      5000
ID                     5000
Kind                   5000
Geo point              5000
Postal code            5000
Public name            5000
Rental status          5000
Scheduled at           5000
Slots                  5000
Station type           5000
Status                 5000
Subscription status    5000
year                   5000
month                  5000
day                    5000
hour                   5000
minute                 5000
dtype: int64

In [ ]:
## is there any dupulcated rows and columns in the table 
df[df.duplicated()]
df


Address  Cars  ...  hour  minute
0                    2 Avenue de Suffren     0  ...    11      43
1              145 Rue Raymond Losserand     6  ...     7      24
2       2 Avenue John Fitzgerald Kennedy     3  ...    20      14
3                 51 Rue Eugène Oudiné     3  ...     4      37
4     6 avenue de la Porte de Champerret     3  ...    17      23
...                                  ...   ...  ...   ...     ...
4995                 8 avenue Ménelotte     2  ...    11      26
4996                   37 rue de Dantzig     4  ...    16      56
4997                      142 rue du Bac     1  ...     7       1
4998         2 avenue du Val de Fontenay     2  ...    17      27
4999                 5 Rue Henry Brisson     4  ...    10       2

[5000 rows x 25 columns]

In [ ]:
# There are no dupulicated rows in our data

In [ ]:
# Am going to check which coluns are useful for this data and which one should be remove 
df.columns 

Index(['Address', 'Cars', 'Bluecar counter', 'Utilib counter',
       'Utilib 1.4 counter', 'Charge Slots', 'Charging Status', 'City',
       'Displayed comment', 'ID', 'Kind', 'Geo point', 'Postal code',
       'Public name', 'Rental status', 'Scheduled at', 'Slots', 'Station type',
       'Status', 'Subscription status', 'year', 'month', 'day', 'hour',
       'minute'],
      dtype='object')

In [ ]:
# some of this columns are not useful for analysing data and they can be removed 
columns_to_drop = ['Charge Slots','Displayed comment','Charging Status','Rental status','Scheduled at', 'Slots','Station type','Subscription status']

# drop this columns 
df.drop(columns_to_drop, inplace=True, axis = 1)


In [ ]:
df.columns

Index(['Address', 'Cars', 'Bluecar counter', 'Utilib counter',
       'Utilib 1.4 counter', 'City', 'ID', 'Kind', 'Geo point', 'Postal code',
       'Public name', 'Status', 'year', 'month', 'day', 'hour', 'minute'],
      dtype='object')

In [49]:
## Trying to make sure that the columns are clean by changing the cases and also removing the space between the worda 
df.columns = df.columns.str.lower().str.strip().str.replace(' ', '_').str.replace('.', '')


In [53]:
 ## To find the cars returns we have to understand what return means, therefore my assumption is that
 ## if the car is picked that it is negative 
 ## if the car is return that it is postive 
 ##zero means no car was picked or returned

 ##creating column for bluecar usage in paris 
df['usage_blue'] = df.bluecar_counter.diff()

# creating column for utilib usage in paris 
df['usage_utilib'] = df.utilib_counter.diff()

# creating column for utilib 14 usage in paris 
df['usage_utilib_14'] = df.utilib_14_counter.diff()

#print the new created columns
df.head()

address  cars  ...  usage_utilib  usage_utilib_14
0                 2 Avenue de Suffren     0  ...           NaN              NaN
1           145 Rue Raymond Losserand     6  ...           0.0              0.0
2    2 Avenue John Fitzgerald Kennedy     3  ...           0.0              2.0
3              51 Rue Eugène Oudiné     3  ...           1.0             -2.0
4  6 avenue de la Porte de Champerret     3  ...          -1.0              0.0

[5 rows x 28 columns]

In [ ]:
# how many Bluecar where used during the interval of 3 days cumulative summation 
df["Bluecar counter"].cumsum()

0           0
1           6
2           9
3          12
4          15
        ...  
4995    11657
4996    11661
4997    11662
4998    11664
4999    11668
Name: Bluecar counter, Length: 5000, dtype: int64

In [ ]:
## How many of bluecars counter were used during the 3 days experiment 
df["Bluecar counter"].sum()

11668

In [54]:
# The data entries where the city is only Paris , there 2816 rows × 25 columns
df[df.city == "Paris"]

address  cars  ...  usage_utilib  usage_utilib_14
0                    2 Avenue de Suffren     0  ...           NaN              NaN
1              145 Rue Raymond Losserand     6  ...           0.0              0.0
3                 51 Rue Eugène Oudiné     3  ...           1.0             -2.0
4     6 avenue de la Porte de Champerret     3  ...          -1.0              0.0
5                   8 Boulevard Voltaire     0  ...           0.0              0.0
...                                  ...   ...  ...           ...              ...
4992                     3 avenue d'Ivry     0  ...           0.0              0.0
4994                    15 rue de Rocroy     0  ...           0.0              0.0
4996                   37 rue de Dantzig     4  ...           0.0              0.0
4997                      142 rue du Bac     1  ...           0.0              0.0
4999                 5 Rue Henry Brisson     4  ...           0.0              0.0

[2816 rows x 28 columns]

The questions to the problem 

most popular hour of the day for picking up a blue share electric car in Paris over the month of April 2018.

In [55]:
### am assinging negative here because the cars are being picked 
df[(df.city== 'Paris') & (df.usage_blue < 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head()

hour
21    80
16    68
7     64
13    62
5     62
Name: hour, dtype: int64

The most popular hour of picking is 9pm with a peak pick of 80 bluecars 

Popular hour for returning the car in Paris 

In [56]:
df[(df.city== 'Paris') & (df.usage_blue > 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head()


hour
3     49
12    48
10    48
6     46
8     44
Name: hour, dtype: int64

Most popular hour for returning blue cars is 03:00 Am

What station is the most popular overall 

In [57]:
## station with the highest number of picked overall
df[(df.kind == 'STATION') & (df.status == 'ok')].groupby('public_name')[['kind']].count().sort_values(by = 'kind', ascending  = False).head()

kind
public_name                     
Paris/Porte de Montrouge/8    13
Paris/Voltaire/182            11
Sèvres/Wolfenbüttel/1       11
Paris/Philippe Auguste/126    11
Paris/Mathis/35               11

The most popular station by public name is Porte De Montrouge with 13 

What station most popular during picking hours 

In [60]:
## The station with the most picking hours 
df[df.usage_blue < 0].groupby('public_name').count()[['usage_blue']].sort_values(by = 'usage_blue', ascending = False).head()

usage_blue
public_name                          
Paris/Grenelle 31/Parking           9
Paris/Flandre/169                   8
Paris/Voltaire/182                  7
Paris/Madame/18                     7
Paris/Bernardins/23                 7

The station popular during picking is Paris/Grenelle 31/Parking	9

Postal code most popular for picking up blue cars

In [64]:
### out puting the most frequency postal code for picking up and the number of car picked at that code 
df[df.usage_blue < 0].groupby('postal_code').count()[['usage_blue']].sort_values(by = 'usage_blue', ascending = False).head()

usage_blue
postal_code            
75015               116
75008               102
75017                96
75012                96
75016                81

The postal code 75015 is the most popular during picking , 116 cars wwere picked from that postal code 

Does results changes if we use Utilib and Utilib 1.4  instead of blue cars

In [65]:
## Using utilib to see if it is more busier than the blue car 
df[(df.city== 'Paris') & (df.usage_utilib > 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head()

hour
6     12
11     9
17     9
8      9
14     8
Name: hour, dtype: int64

In [66]:
## Using utilib 1.4 to see if it is more busier than the blue car 
df.city== 'Paris') & (df.usage_utilib_14 > 0)].groupby('hour')['hour'].count().sort_values(ascending = False).head()

hour
9     17
17    15
5     15
21    14
22    14
Name: hour, dtype: int64

If we would have used Utilib and Utilib 1.4 instead of blue cars there will have been no changes because neither the Ulilib or Utilib 1.4 is as busy as BlueCar.